packages/salesforce-adapter/src/dependency_changer.ts

```diff
@@ -0,0 +1,77 @@
+/*
+*                      Copyright 2024 Salto Labs Ltd.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with
+* the License.  You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+import {
+  DependencyChanger, isInstanceChange, getChangeData, ElemID, Change, InstanceElement, DependencyChange, ChangeId,
+  isAdditionOrModificationChange,
+} from '@salto-io/adapter-api'
+import { collections, values } from '@salto-io/lowerdash'
+import { isInstanceOfCustomObjectSync } from './filters/utils'
+import { isCustomObject } from './transformers/transformer'
+
+const { awu } = collections.asynciterable
+const { isDefined } = values
+
+const generateInstanceToTypeDep = (
+  changeId: ChangeId,
+  instance: InstanceElement,
+  typeToChangeIdMap: Map<ElemID, ChangeId>
+): DependencyChange | undefined => {
+  const type = instance.getTypeSync()
+  const typeChangeId = typeToChangeIdMap.get(type.elemID)
+  if (typeChangeId === undefined) {
+    return undefined
+  }
+  return {
+    action: 'add',
+    dependency: {
+      source: changeId,
+      target: typeChangeId,
+    },
+  }
+}
+
+const dataRecordToAssociatedType: DependencyChanger = async changes => {
+  // Note that we don't handle removal yet. We should probably create a reverse dependency to ensure we delete all
+  // records before we delete their type.
+  const instanceChanges: [ChangeId, Change<InstanceElement>][] = Array.from(changes.entries())
+    .filter(([, change]) => isInstanceChange(change))
+    .filter(([, change]) => isAdditionOrModificationChange((change)))
+    .filter(([, change]) => (
+      isInstanceOfCustomObjectSync(getChangeData(change))
+    )) as [ChangeId, Change<InstanceElement>][]
+  const typeChanges = Array.from(changes.entries())
+    .filter(([, change]) => isCustomObject(getChangeData(change)))
```

Comment:
```suggestion
    .filter(([, change]) => isCustomObjectSync(getChangeData(change)))
```
This statement is always true...

Please make the UTs more robust so they would catch this edge-case. Considering they don't contain any other change but a single change on a type, they didn't catch this.

Reply:
So, funny story - this is actually (accidentally) not a bug, because this logic maps instances to types, and we filter the instances correctly. So there will be extra entries in `typeToChangeIdMap`, but no instance would map to them (because we only map instances of custom objects), and everything will come out OK. 
I will still fix this, of course, but I thought it's interesting.